### PR TITLE
Print warning when a resource name is ambiguous

### PR DESCRIPTION
**Pull request by JoelSpeed**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

When multiple CRDs are installed within a cluster, and a user uses a non-groupified name to look up the resource, eg `kubectl get foo` or `kubectl get foos` when we have `foo.bar.baz` and `foo.whoosh.bang` as CRDs, they may not get the expected result.

When kubectl looks up resources, it has a number of rules for prioritising which of the two CRDs would win in this case, for example the group version can affect the decision, as well as the name (alphabetical sorting).

So we have seen examples where `kubectl get <resource>` can vary, cluster to cluster based on the installed CRDs.

While I know users should use groupified names to solve this, eg `kubectl get foo.bar.baz` or `kubectl get foo.bar` would work, there's no obvious warning that the user has put an ambiguous resource or kind name.

This PR aims to solve that by adding a warning, so that if the arg they pass matches multiple resources, they are informed.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

The resource package is imported by the genericclioptions package, hence we cannot pass the entire IOStreams structure into `WithWarningPrinter`.

#### Does this PR introduce a user-facing change?

```release-note
Kubectl will now warn users when short type names match multiple registered resources
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/cli-runtime/pkg/resource/builder.go [hunk 1]
```diff
@@ -799,7 +833,11 @@ func (b *Builder) mappingFor(resourceOrKindArg string) (*meta.RESTMapping, error
 		return nil, err
 	}
 
-	return mapping, nil
+	if len(mappings) > 1 {
+		b.printWarning(resourceAmbiguousFmt, resourceOrKindArg)
```
Comment (deads2k) [inline staging/src/k8s.io/cli-runtime/pkg/resource/builder.go:837]: I haven't reviewed to be sure this printing technique is in keeping with printing warnings sent back from the kube-apiserver, but the idea of printing a warning the same way for ambiguous restmappings is good.

**Comment by BenTheElder** [comment]
xref: https://github.com/kubernetes/kubernetes/issues/102378, https://github.com/kubernetes/kubernetes/issues/108573

File: staging/src/k8s.io/cli-runtime/pkg/resource/builder.go (hunk 1 above)
Comment (JoelSpeed) [reply staging/src/k8s.io/cli-runtime/pkg/resource/builder.go:837]: This uses the new warning printer which was added in #108210, the goal of that PR is to start making alignments for the warnings printed by kubectl so I believe this is the best route to take, whether that aligns with API warnings or not I'm not 100% sure

**Comment by JoelSpeed** [comment]
Thanks for the suggestion @sftim!

/hold

This was discussed a couple of weeks ago on the sig-cli call. I have some investigative work to do and some questions to answer for the group before I come back to this. Will try to get to that soon

**Comment by deads2k** [comment]
> This was discussed a couple of weeks ago on the sig-cli call. I have some investigative work to do and some questions to answer for the group before I come back to this. Will try to get to that soon

Just making sure this doesn't fall off the radar.  This PR looks quite useful.

**Comment by JoelSpeed** [comment]
> Just making sure this doesn't fall off the radar. This PR looks quite useful.

Apologies for the delays, this is very much still on my radar and I'm intending to delve back into it over the next week or two

**Comment by JoelSpeed** [comment]
This PR was discussed on the [June 1st SIG-CLI](https://docs.google.com/document/d/1r0YElcXt6G5mOWxwZiXgGu_X6he3F--wKwg-9UBc29I/edit#heading=h.1x468k30seiw) call. I wanted to capture the conversation here to get an idea for how to proceed. There are some open questions which I'll come back later to address.

Some confusing issues with the end user experience which we would like to alleviate:
* Different resources are prioritized by Kubectl depending on a few factors, the priority (not yet configurable for CRD?), the group version, then the name (alphabetical)
* Scenario: If I have two CRDs, `foo.v1beta1.bar` and `foo.v1beta1.baz` (same resource and version) installed in my cluster 
  * Because these are the same version, `kubectl get foo` returns `foo.v1beta1.bar` (because it is alphabetically first).
  * If then someone creates a new CRD, `bash.v1.baz`, this means the `baz` group priority is now higher than the bar group priority (because `v1` is prioritised over `v1beta1`, suddenly, even though nothing changed for any `foo` resource, `kubectl get foo` returns `foo.v1beta1.baz`
  * This instability is something we should definitely try to fix, changing the priority because something else in the group was installed at a more stable version is very confusing
* Kubectl also does prefix matching, eg `foo.group1.baz` and `foo.group1.bar` are both valid resources, if only one is installed, `kubectl get foo.group1` could return the correct item even though it’s not fully qualified, if both are installed we revert to prior issues

A summary of the comments made:
* We want to solve this issue because it's easy for users to hit this
* We don't want to see this all the time, eg if I got a warning every time I did `kubectl get po` I'd probably be annoyed even though I know it's wrong
* We want to teach our users to use the fully-groupified names
* This seems like it could be quite a personal decision
* We may be able to solve the instability issue separately to improve UX there

Questions that were raised during the call:
* What happens when an API is moved between groups, would this warn on built-ins that moved groups (note API machinery recommend against this now)?
  * Answer: Yes, it would likely warn if they are CRDs (need to check built ins) but this is something API machinery are forbidding going forward for built-ins so this would only happen for CRDs going forward, and as there's no priority ordering for these, there's no way to select server side, which should be preferred.
* Does this effect users when a version is upgraded?
  * Answer: No. The logic doesn't take into account the version of individual resources. The confusing behaviour only happens if there is an overlapping `Resource` within multiple `Groups`. For the same resource in a `GroupResource`, multiple versions are handled within the CRD and so Kubectl just gets the preferred version to deal with.
* Is this a cluster admin issue? Should they be installing multiple CRDs with similar short names? They're the ones controlling the alias
  * Answer: Yes and no, they could configure the aliases but the short names are likely fixed. It is valid that many third party operators could create their own CRDs with similar short names (which is where this issue comes in). Is it an error for an admin to install multiple community operators with similar short names, probably not?

Open Questions:
* How would this affect built-ins/is there a way to limit this so that it only happens when a CRD is present?
* When there's an ambiguity, can we let users resolve the issue and remember their choice? Perhaps an rc style file could store preferences for the user

CC @soltysh @KnVerey I believe you two were both interested in this, apologies for the delay in writing this up but do you recall any other points we need to clarify here? (I've gone through the call recording to make these notes so I think I captured everything)

**Comment by Ritikaa96** [comment]
Hi @JoelSpeed , the same issue is faced in OpenShift with volumesnapshot provided by CSI and virtualservices provided by ServiceMesh. This PR is extremely useful. Are we waiting for further addition before merging it?

**Comment by JoelSpeed** [comment]
@Ritikaa96 There are a couple of open questions from my previous comment that still need to be addressed, I haven't had a chance to look into how this would be done yet. But if others are willing to continue the discussion then I welcome that. Else I'll try and find some time to get to this soon

**Comment by Ritikaa96** [comment]
> But if others are willing to continue the discussion then I welcome that. Else I'll try and find some time to get to this soon

@JoelSpeed thanks for the reply. I can see from your previous comments


> Open Questions:
> 
> * How would this affect built-ins/is there a way to limit this so that it only happens when a CRD is present?
> * When there's an ambiguity, can we let users resolve the issue and remember their choice? Perhaps an rc style file could store preferences for the user
> 
> CC @soltysh @KnVerey I believe you two were both interested in this, apologies for the delay in writing this up but do you recall any other points we need to clarify here? (I've gone through the call recording to make these notes so I think I captured everything)

Are these what you are talking about?

Looking forward to this one, appreciate the work! :)

File: staging/src/k8s.io/cli-runtime/go.mod [hunk 2]
```diff
@@ -19,13 +19,15 @@ require (
 	k8s.io/apimachinery v0.0.0
 	k8s.io/client-go v0.0.0
 	k8s.io/kube-openapi v0.0.0-20220328201542-3ee0da9b0b42
+	k8s.io/kubectl v0.0.0
```
Comment (liggitt) [inline staging/src/k8s.io/cli-runtime/go.mod:22]: as an aside, this is adding a dependency cycle between k8s.io/kubectl <-> k8s.io/cli-runtime ... the code should be structured so that we don't have cycles in our modules

Comment (JoelSpeed) [reply staging/src/k8s.io/cli-runtime/go.mod:22]: Oops, good catch, I'm intending to come back to this PR next week, I'll look into this as well, thanks

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

**Comment by JoelSpeed** [comment]
> If you still need this PR then please rebase, if not, please close the PR

Apologies, this is still relevant, I just haven't found the time to get to it, hoping to look at it later this week

File: staging/src/k8s.io/cli-runtime/go.mod (hunk 2 above)
Comment (JoelSpeed) [reply staging/src/k8s.io/cli-runtime/go.mod]: @liggitt I've just pushed a new version of this where I've moved the code needed from kubectl into cli-runtime and then put deprecated aliases in place of the old functions, does that seem like an appropriate way to solve the import cycle? That's now the first commit in the branch as a prefactor step

**Comment by sftim** [comment]
There are CRDs by the same name as built-in APIs. Some Ingress controllers deliberately name their custom resource kind `Ingress` (which I'm not keen on, but it happens). And yep, no warning.

Other plausible places to see a mixup:
- `events`
- `services`

**Comment by JoelSpeed** [comment]
> If that's the existing behaviour, then I see even more reason to add something like this warning to protect users in case a CRD does start overriding a built-in with the current Kubectl logic

After some more digging, it would seem that, kubectl uses the standard deferred discovery rest mapper, this in turn uses the API server to discover what API groups exists and prioritises them based on the group order they are returned by the API server.

The API server returns all built in groups before it starts returning CRD groups, CRD groups are alphabetically returned it seems. I'm guessing this is based on group priority? CRD priorities are low and can't be configured IIRC.

So, no, it is not possible for someone to overthrow a built in type with a CRD, as this will always appear later in the discovered list than a built-in type.

That said, kubectl has no way to know what is built in and what isn't, this isn't information that is returned by the API server (AFAICT), or at least, it's not data that makes its way into the rest mapper.

So to the question of would this affect built-ins, yes, if someone created a new CRD which duplicates the name of some built in type, then the built-in would be returned first, but they may have been wanting their CRD kind, so perhaps a warning is still valid?

---

> When there's an ambiguity, can we let users resolve the issue and remember their choice? Perhaps an rc style file could store preferences for the user

Moving onto this second question, this feels like a new feature that would need further discussion, perhaps even a KEP, I think discussing this with SIG-CLI before proceeding would be a better choice. But I agree, having some way to silence the warnings would be nice, having a way to choose a preferred API group when the result is ambiguous, would also be nice.

At the moment, AFAIK there's no configuration of kubectl at all, does anyone know of any prior art/keps around the idea of a Kubectl config, what are the other features that might be included here?

File: staging/src/k8s.io/cli-runtime/pkg/printers/terminal.go [hunk 3]
```diff
@@ -37,3 +41,35 @@ func WriteEscaped(writer io.Writer, output string) error {
 func EscapeTerminal(in string) string {
 	return terminalEscaper.Replace(in)
 }
+
+// IsTerminal returns whether the passed object is a terminal or not
+func IsTerminal(i interface{}) bool {
```
Comment (ardaguclu) [inline staging/src/k8s.io/cli-runtime/pkg/printers/terminal.go:46]: I think we can do this movement in separate PR as a preparation of this PR.

Comment (ardaguclu) [reply staging/src/k8s.io/cli-runtime/pkg/printers/terminal.go:46]: This commit already looks good to me. But if you don't have time, I can also take over it and open PR.

**Comment by ardaguclu** [comment]
> So to the question of would this affect built-ins, yes, if someone created a new CRD which duplicates the name of some built in type, then the built-in would be returned first, but they may have been wanting their CRD kind, so perhaps a warning is still valid?

builtin(`po`), CRD(`shortname: po`) conflict:
It is true that always builtin type is returned in case of a conflict with a CRD. But wouldn't it be better to warn user about the CRD that is conflicting with builtin type. That forces users to use fully qualified names(which we always recommend). That also increases awareness of the incorrect shortnaming of CRD and people can rise issues to the CRD owner to fix it. And lastly, users will know that when they use `po`, their CRDs will not be picked. 

CRD1(`shortname: foo`), CRD2(`shortname: foo`):
Currently, non-deterministic behavior of which one will be picked is really problematic and dynamic nature of CRDs make it more difficult to overcome(I have CRD1 and now after I installed CRD2, `foo` starts returning CRD2). We should definitely show a warning about which one is picked and the which ones are discarded.

And this PR solves both of the issues above and we can merge it with a few refactorings(Since we are raising this warning using the warning printer, other tools rely on `builder` can easily mute this warning.).

File: staging/src/k8s.io/cli-runtime/pkg/resource/builder.go [hunk 4]
```diff
@@ -768,7 +795,13 @@ func (b *Builder) mappingFor(resourceOrKindArg string) (*meta.RESTMapping, error
 		gvk, _ = restMapper.KindFor(*fullySpecifiedGVR)
 	}
 	if gvk.Empty() {
-		gvk, _ = restMapper.KindFor(groupResource.WithVersion(""))
+		gvks, _ := restMapper.KindsFor(groupResource.WithVersion(""))
```
Comment (ardaguclu) [inline staging/src/k8s.io/cli-runtime/pkg/resource/builder.go:798]: Although this PR looks promising, unfortunately `KindsFor` does not return the list of resources belonging to the given short name. Because `KindsFor` calls `expandResourceShortcut` in https://github.com/kubernetes/kubernetes/blob/9924dc65b75975c5993da95ea8e0cb9360543c3f/staging/src/k8s.io/client-go/restmapper/shortcut.go#L62 and this function https://github.com/kubernetes/kubernetes/blob/9924dc65b75975c5993da95ea8e0cb9360543c3f/staging/src/k8s.io/client-go/restmapper/shortcut.go#L155 already picks the one of the resources instead the list of resources with given short name.

**Comment by ardaguclu** [comment]
After digging into this issue further, I proposed this https://github.com/kubernetes/kubernetes/pull/117668

**Comment by ardaguclu** [comment]
> I think the plan is to close this in favour of #117668, right @ardaguclu ?

Yes, that is correct.

**Comment by JoelSpeed** [comment]
Closing in favour of #117668
